test_scripts/Policies/Policy_Table_Update/177_PTU_is_performed_for_several_Apps_if_HMI_sends_BC_OnSystemRequest_with_url_parameter.lua

```diff
@@ -0,0 +1,80 @@
+---------------------------------------------------------------------------------------------------
+-- Issues: https://github.com/smartdevicelink/sdl_core/issues/3715,
+--         https://github.com/smartdevicelink/sdl_core/issues/3881
+---------------------------------------------------------------------------------------------------
+-- Description:
+--  SDL accepts PTU from application for several Apps in case HMI sends BC.OnSystemRequest with 'url' parameter
+-- Steps:
+-- 1. Core and HMI are started
+-- 2. App1 is registered
+-- 3. App2 is registered
+-- 4. Core sends OnStatusUpdate UPDATING
+-- 5. HMI sends OnSystemRequest with 'url' parameter
+-- 6. After OnSystemRequest is sent to randomly app, app attempts to send SystemRequest during PTU
+-- SDL does:
+--  - accept the PTU, responding success:true, resultCode:SUCCESS
+---------------------------------------------------------------------------------------------------
+
+--[[ Required Shared libraries ]]
+local utils = require('user_modules/utils')
+local runner = require('user_modules/script_runner')
+local common = require("user_modules/sequences/actions")
+
+--[[ Test Configuration ]]
+runner.testSettings.isSelfIncluded = false
+runner.testSettings.restrictions.sdlBuildOptions = { { extendedPolicy = { "PROPRIETARY", "EXTERNAL_PROPRIETARY" } } }
+
+--[[ Local Functions ]]
+local function policyTableUpdate()
+  local ptuFileName = os.tmpname()
+  local requestId = common.hmi.getConnection():SendRequest("SDL.GetPolicyConfigurationData",
+      { policyType = "module_config", property = "endpoints" })
+  common.hmi.getConnection():ExpectResponse(requestId)
+  :Do(function()
+      common.hmi.getConnection():SendNotification("BasicCommunication.OnSystemRequest", {
+        requestType = "PROPRIETARY",
+        fileName = common.sdl.getPTSFilePath(),
+        url = "http://x.x.x.x:3000/api/1/policies/proprietary"
+      })
+      local ptuTable = common.getPTUFromPTS()
+      for i, _ in pairs(common.mobile.getApps()) do
```

Comment:
This test fails sometimes with the following error in the console logs
```
[09:52:23,032] Successful_PTU_via_a_mobile_app                                                       [FAIL] (12400 ms)
 Response to 3: ValidIf: 
payload.resultCode: expected: SUCCESS, actual value: REJECTED
 HMI call BasicCommunication.SystemRequest: Timeout: HMI call BasicCommunication.SystemRequest: Timeout expired
```
and `PolicyHandler::IsPTUSystemRequestAllowed: PTU received from unexpected application, request was sent to 131073` in the core logs.

Since one app is chosen randomly for the PTU (https://github.com/smartdevicelink/sdl_core/pull/3883/files#diff-c90310eb34a277f2159d2f0588fddce48cb3235f0e8557017d2a988aad562da2R90) I'm not sure if we should loop through all registered apps with the same expected events.

Reply:
@ShobhitAd I believe there is an SDL issue smartdevicelink/sdl_core#3881 (comment) Since one app is chosen randomly for the PTU we define App for PTU through loop for all registered apps and :Times(AtMost(1))

Reply:
```
local function policyTableUpdate()
  local ptuFileName = os.tmpname()
  local requestId = common.hmi.getConnection():SendRequest("SDL.GetPolicyConfigurationData",
      { policyType = "module_config", property = "endpoints" })
  common.hmi.getConnection():ExpectResponse(requestId)
  :Do(function()
      common.hmi.getConnection():SendNotification("BasicCommunication.OnSystemRequest", {
        requestType = "PROPRIETARY",
        fileName = common.sdl.getPTSFilePath(),
        url = "http://x.x.x.x:3000/api/1/policies/proprietary",
        appID = common.getHMIAppId()
      })
      local ptuTable = common.getPTUFromPTS()
      for i, _ in pairs(common.mobile.getApps()) do
        ptuTable.policy_table.app_policies[common.app.getPolicyAppId(i)] = common.ptu.getAppData(i)
      end
      utils.tableToJsonFile(ptuTable, ptuFileName)
      local event = common.run.createEvent()
      common.hmi.getConnection():ExpectEvent(event, "PTU event")

      common.mobile.getSession(id):ExpectNotification("OnSystemRequest", { requestType = "PROPRIETARY" })
      :Do(function()
          common.hmi.getConnection():ExpectRequest("BasicCommunication.SystemRequest")
          :Do(function(_, d3)
              common.hmi.getConnection():ExpectRequest("VehicleInfo.GetVehicleData", { odometer = true })
              common.hmi.getConnection():ExpectNotification("SDL.OnStatusUpdate", { status = "UP_TO_DATE" })
              common.hmi.getConnection():SendResponse(d3.id, "BasicCommunication.SystemRequest", "SUCCESS", { })
              common.hmi.getConnection():SendNotification("SDL.OnReceivedPolicyUpdate",
                { policyfile = d3.params.fileName })
            end)
          common.hmi.getConnection():RaiseEvent(event, "PTU event")
          local corIdSystemRequest = common.mobile.getSession(id):SendRPC("SystemRequest", {
            requestType = "PROPRIETARY" }, ptuFileName)
          common.mobile.getSession():ExpectResponse(corIdSystemRequest, { success = true, resultCode = "SUCCESS" })
          :Do(function() os.remove(ptuFileName) end)
        end)
      :Times(AtMost(1))

    end)
end
```

Reply:
@VjKlepikov I think the failure is expected. Because SDL Core is randomly choosing an app, if the app sending the system request did not receive the OnSystemRequest Core will reject the request. This behavior was implemented via this proposal: https://github.com/smartdevicelink/sdl_evolution/blob/master/proposals/0337-reject-proprietary-http-systemrequests-when-ptu-not-in-progress.md

Reply:
@JackLivio As I can see in ATF log , the app2 is received OnSystemRequest from Core

===== Successful_PTU_via_a_mobile_app : 
...
SDL->MOB [24-03-2022 15:53:40,604] [rpcFunction: OnSystemRequest, sessionId: 2, version: 3, frameType: 1, encryption: false, serviceType: 15, frameInfo: 0, messageId: 6, binaryDataSize: 21426] : {"fileType":"JSON","url":"http:\/\/x.x.x.x:3000\/api\/1\/policies\/proprietary","requestType":"PROPRIETARY"} 

MOB->SDL [24-03-2022 15:53:40,606] [rpcFunction: SystemRequest, sessionId: 2, version: 3, frameType: 1, encryption: false, serviceType: 7, frameInfo: 0, messageId: 5, binaryDataSize: 17338] : {"requestType":"PROPRIETARY"} 

SDL->MOB [24-03-2022 15:53:40,615] [rpcFunction: SystemRequest, sessionId: 2, version: 3, frameType: 1, encryption: false, serviceType: 7, frameInfo: 0, messageId: 7, binaryDataSize: 0] : {"success":false,"resultCode":"REJECTED"}

Reply:
@VjKlepikov Ok thank you for the clarification, investigating this issue